jax/interpreters/partial_eval.py

```diff
@@ -342,10 +344,14 @@ def abstract_eval_fun(fun, *avals, **params):
   return avals_out
 
 
+Recipe = Union['JaxprEqnRecipe', 'LambdaBinding', 'FreeVar', 'ConstVar',
```

Comment:
I found this definition surprising, with so many different things mixed up in there. It seems that this is used in only one place, as the optional `recipe` arg for JaxprTracer.__init__? 

Perhaps a more specific name, such as `JaxprTracerRecipe` is better? Or drop it altogether?

Reply:
I think the first four entries of this list are all recipes. `Literal` could be distinguished from `LiteralRecipe`, though I was lazy and just reused the same type. Having `core.Unit` in there could also be revised. Other than those two exceptions, though, I think these are all arms of a conceptual "recipe" type.

Reply:
Renamed to `JaxprTracerRecipe`.

The reason for the annotation is to codify what things may appear in this function position. It's not so much that I want to introduce a name, more that I don't want to write a `Union` with 6 arms inline in a function definition.